File: src/domain/reward.ts
export interface RewardContribution {
  agentId: string;
  contributionScore: number;
}

export interface RewardAllocation {
  agentId: string;
  amountLamports: number;
}

export interface RewardResult {
  allocations: RewardAllocation[];
  retainedLamports: number;
}

const MINIMUM_ACCEPTED_SCORE = 60;

export function calculateVirtualReward(
  rewardPoolLamports: number,
  finalScore: number,
  contributions: RewardContribution[],
): RewardResult {
  if (!Number.isSafeInteger(rewardPoolLamports) || rewardPoolLamports < 0) {
    throw new Error("rewardPoolLamports must be a non-negative safe integer");
  }
  if (finalScore < MINIMUM_ACCEPTED_SCORE || finalScore > 100) {
    throw new Error("finalScore must be between 60 and 100 for Reward allocation");
  }
  if (contributions.length === 0) throw new Error("at least one contribution is required");
  if (contributions.some(({ contributionScore }) => contributionScore <= 0 || contributionScore > 100)) {
    throw new Error("contributionScore must be between 1 and 100");
  }

  const payable = Math.floor((rewardPoolLamports * finalScore) / 100);
  const totalWeight = contributions.reduce((sum, item) => sum + item.contributionScore, 0);
  let allocated = 0;

  const allocations = contributions.map((item, index) => {
    const amountLamports =
      index === contributions.length - 1
        ? payable - allocated
        : Math.floor((payable * item.contributionScore) / totalWeight);
    allocated += amountLamports;
    return { agentId: item.agentId, amountLamports };
  });

  return { allocations, retainedLamports: rewardPoolLamports - payable };
}

File: services/rcao/app/reward.py
from .models import RewardAllocation, RewardContribution, RewardResult


MINIMUM_ACCEPTED_SCORE = 60


def calculate_virtual_reward(
    reward_pool_lamports: int,
    final_score: int,
    contributions: list[RewardContribution],
) -> RewardResult:
    if reward_pool_lamports < 0:
        raise ValueError("reward_pool_lamports must be non-negative")
    if final_score < MINIMUM_ACCEPTED_SCORE or final_score > 100:
        raise ValueError("final_score must be between 60 and 100")
    if not contributions:
        raise ValueError("at least one contribution is required")

    payable = (reward_pool_lamports * final_score) // 100
    total_weight = sum(item.contribution_score for item in contributions)
    allocated = 0
    allocations: list[RewardAllocation] = []

    for index, contribution in enumerate(contributions):
        amount = (
            payable - allocated
            if index == len(contributions) - 1
            else (payable * contribution.contribution_score) // total_weight
        )
        allocated += amount
        allocations.append(
            RewardAllocation(
                agent_id=contribution.agent_id,
                amount_lamports=amount,
            )
        )

    return RewardResult(
        allocations=allocations,
        retained_lamports=reward_pool_lamports - payable,
    )

